evm/db/journal.py

```diff
@@ -1,226 +1,258 @@
 
-        return self.wrapped_db.delete(key)
+    def delete(self, key: bytes) -> None:
+        if key not in self.journal and key not in self.wrapped_db:
+            raise KeyError(key)
+        del self.journal[key]
 
     #
     # Snapshot API
     #
-    def _validate_checkpoint(self, checkpoint):
+    def _validate_changeset(self, changeset_id: uuid.UUID) -> None:
         """
-        Checks to be sure the checkpoint is known by the journal
+        Checks to be sure the changeset is known by the journal
         """
-        if checkpoint not in self.journal:
-            raise ValidationError("Checkpoint not found in journal: {0}".format(
-                str(checkpoint)
+        if not self.journal.has_changeset(changeset_id):
+            raise ValidationError("Changeset not found in journal: {0}".format(
+                str(changeset_id)
             ))
 
-    def snapshot(self):
+    def record(self) -> uuid.UUID:
         """
-        Takes a snapshot of the database by creating a checkpoint.
+        Starts a new recording and returns an id for the associated changeset
         """
-        return self.journal.create_checkpoint()
+        return self.journal.record_changeset()
 
-    def revert(self, checkpoint):
+    def discard(self, changeset_id: uuid.UUID) -> None:
         """
-        Reverts the database back to the checkpoint.
+        Throws away all journaled data starting at the given changeset
         """
-        self._validate_checkpoint(checkpoint)
+        self._validate_changeset(changeset_id)
+        self.journal.pop_changeset(changeset_id)
 
-        for key, value in self.journal.pop_checkpoint(checkpoint).items():
-            if value is None:
-                self.wrapped_db.delete(key)
-            else:
-                self.wrapped_db.set(key, value)
+    def commit(self, changeset_id: uuid.UUID) -> None:
+        """
+        Commits a given changeset. This merges the given changeset and all
+        subsequent changesets into the previous changeset giving precidence
+        to later changesets in case of any conflicting keys.
 
-    def commit(self, checkpoint):
+        If this is the base changeset then all changes will be written to
+        the underlying database and the Journal starts a new recording.
         """
-        Commits a given checkpoint.
+        self._validate_changeset(changeset_id)
+        journal_data = self.journal.commit_changeset(changeset_id)
+
+        if self.journal.is_empty():
+            for key, value in journal_data.items():
+                if value is not DELETED_ENTRY:
+                    self.wrapped_db[key] = value
+                else:
+                    try:
+                        del self.wrapped_db[key]
+                    except KeyError:
+                        pass
+
+            # Ensure the journal automatically restarts recording after
+            # it has been persisted to the underlying db
+            self.record()
```

Comment:
We can use `self.record()` here because we *implicitely* know that the journal is empty.  However, do you think it'd be better to use `reset()`?  That'd remove any possibility of previous journal state hanging around.

Suppose a future refactor the the `Journal` adds some additional container where state is stored... Then this might not do what we want.

Reply:
Personally, I'd rather not call `reset()` to not create lots of instances that need to be garbage collected. Your call.

Reply:
Since this should only occur once per transaction I doubt the perf hit will be measurable.